packages/functional-tests/tests/misc/keyStretchingV2.spec.ts

```diff
@@ -0,0 +1,660 @@
+import { RecoveryKeyPage } from '../../pages/settings/recoveryKey';
+import { ChangePasswordPage } from '../../pages/settings/changePassword';
+
+/**
+ * These tests represent various permutations between interacting with V1 and V2
+ * key stretched passwords. We need to ensure that operations are interchangeable!
+ */
+test.describe('key-stretching-v2', () => {
+  let email = '';
+  let password = '';
+
+  function getClient(url: string, version: SaltVersion) {
+    return new AuthClient(url, {
+      keyStretchVersion: version,
+    });
+  }
+
+  test.beforeEach(() => {
+    test.slow();
+    email = `signin${crypto.randomBytes(8).toString('hex')}@restmail.net`;
+    password = `${crypto.randomBytes(10).toString('hex')}`;
+  });
+
+  /**
+   * For full coverage, we have to run many flows on both react, and backbone version.
+   * There are differences between how react and backbone use authClient and gql-api
+   * for these types of operations.
+   */
+  ['react', 'backbone'].forEach((mode) => {
+    /**
+     * Complete set of types used by tests and helper functions.
+     */
+    type Opts = {
+      page: Page;
+      target: BaseTarget;
+      signupReact: SignupReactPage;
+      login: LoginPage;
+      settings: SettingsPage;
+      resetPassword: ResetPasswordPage;
+      resetPasswordReact: ResetPasswordReactPage;
+      changePassword: ChangePasswordPage;
+      recoveryKey: RecoveryKeyPage;
+    };
+
+    // Helpers
+    async function _checkCredentials(version, opts: Pick<Opts, 'target'>) {
+      const { target } = opts;
+      const client = getClient(target.authServerUrl, 2);
+      if (version === 1) {
+        const status = await client.getCredentialStatusV2(email);
+        expect(status?.clientSalt).toBeUndefined();
+        expect(status.currentVersion).toEqual('v1');
+        expect(status.upgradeNeeded).toBeTruthy();
+      } else if (version === 2) {
+        const status = await client.getCredentialStatusV2(email);
+        expect(status?.clientSalt).toMatch(/quickStretchV2:/);
+        expect(status?.currentVersion).toEqual('v2');
+        expect(status?.upgradeNeeded).toBeFalsy();
+      }
+    }
+
+    async function _getKeys(version, opts: Pick<Opts, 'target'>) {
+      const client = getClient(opts.target.authServerUrl, version);
+      const response = await client.signIn(email, password, { keys: true });
+      const keys = client.accountKeys(
+        response.keyFetchToken,
+        response.unwrapBKey
+      );
+      return keys;
+    }
+
+    async function _signUp(
+      version: 1 | 2,
+      opts: Pick<
+        Opts,
+        'page' | 'target' | 'signupReact' | 'login' | 'settings'
+      >,
+      signOut: boolean
+    ) {
+      const { page, target, signupReact, login, settings } = opts;
+      const stretch = version === 2 ? 'stretch=2' : '';
+
+      if (mode === 'react' || mode === 'backbone') {
+        // singup is at 100% now
+        page.goto(
```

Comment:
Q: Does this need an `await` (like `await page.goto(...)`)?

Reply:
Doesn't hurt...